### PR TITLE
Timeout when waiting for Cachito requests based on the last update time

### PR DESCRIPTION
**Pull request by mprahl**
This will reduce the amount of times that OSBS will fail the build
when Cachito has a long queue of tasks.

Additionally, since the burst length resets after every state
change, the default `burst_retry` value is set from `1` second
to `3` seconds to reduce the load on Cachito's REST API and database.

Resolves CLOUDBLD-2007.

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [ ] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates
- [ ] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/utils/cachito.py [hunk 1]
```diff
@@ -101,7 +102,7 @@ def request_sources(self, repo, ref, flags=None, pkg_managers=None, user=None,
         return response_json
 
     def wait_for_request(
-            self, request, burst_retry=1, burst_length=30, slow_retry=10):
+            self, request, burst_retry=5, burst_length=30, slow_retry=10):
```
Comment (athos-ribeiro) [inline atomic_reactor/utils/cachito.py:105]: Why is this also being changed? Should this change be in a different PR or also be explained in the commit message?

File: atomic_reactor/utils/cachito.py [hunk 2]
```diff
@@ -147,14 +147,16 @@ def wait_for_request(
 
             # All other states are expected to be transient and are not checked.
 
-            elapsed = time.time() - start_time
+            updated_time = datetime.fromisoformat(response_json["updated"])
+            elapsed = (datetime.utcnow() - updated_time).total_seconds()
```
Comment (athos-ribeiro) [inline atomic_reactor/utils/cachito.py:151]: Should we trust the time on an external service here? We could make atomic-reactor start counting from the last time it has perceived  a new update instead.

File: atomic_reactor/utils/cachito.py [hunk 3]
```diff
@@ -147,14 +147,16 @@ def wait_for_request(
 
             # All other states are expected to be transient and are not checked.
 
-            elapsed = time.time() - start_time
+            updated_time = datetime.fromisoformat(response_json["updated"])
```
Comment (athos-ribeiro) [inline atomic_reactor/utils/cachito.py:150]: While `fromisoformat` was introduced in python 3.7, atomic-reactor still supports python 2.7 (and it would also be nice to keep 3.6 compatibility since we plan on supporting CentOS 8 for buildroots soon).

File: atomic_reactor/utils/cachito.py (hunk 1 above)
Comment (mprahl) [reply atomic_reactor/utils/cachito.py]: I added an explanation in the commit message and PR description. I also changed this to `3` seconds.

File: atomic_reactor/utils/cachito.py (hunk 3 above)
Comment (mprahl) [reply atomic_reactor/utils/cachito.py]: Thanks for catching that. I changed it to use `strptime` instead.

File: atomic_reactor/utils/cachito.py (hunk 2 above)
Comment (mprahl) [reply atomic_reactor/utils/cachito.py:151]: That is interesting to consider, however, since Cachito uses Kerberos for its workers to authenticate with the REST API, there would be a Cachito outage if there was more than a 5 minute time gap.

I'll add an extra commit which implements your suggested approach. Then you and other reviewers can let me know which approach you prefer.

Comment (mprahl) [reply atomic_reactor/utils/cachito.py]: @athos-ribeiro I ended up just amending the original commit since this seemed pretty clean.

File: atomic_reactor/utils/cachito.py [hunk 4]
```diff
@@ -147,14 +149,22 @@ def wait_for_request(
 
             # All other states are expected to be transient and are not checked.
 
-            elapsed = time.time() - start_time
+            # If "last_updated_value" does not match the "updated" value of the
+            # request from Cachito, then we know Cachito has performed some work
+            # since the last check, so the timer resets.
+            if last_updated_value is None or last_updated_value != response_json['updated']:
```
Comment (mprahl) [inline atomic_reactor/utils/cachito.py:155]: I added `last_updated_value is None` to the if statement in the event Cachito has a bug and returns a null `updated` value. Let me know if you think this is too defensive and I can remove it.

Comment (lcarva) [reply atomic_reactor/utils/cachito.py:154]: You could simply do:
```
last_update_time = response_json['updated'] or last_update_time or time.time()
```

However, in either case, if there's a Cachito bug this could run indefinitely.

File: atomic_reactor/utils/cachito.py [hunk 5]
```diff
@@ -147,14 +148,22 @@ def wait_for_request(
...
+                last_updated_value = response_json['updated']
+                last_update_time = time.time()
+
+            elapsed = time.time() - last_update_time
```
Comment (lcarva) [inline atomic_reactor/utils/cachito.py:158]: This will cause a failure if Cachito doesn't return an `updated` value initially.

Comment (mprahl) [reply atomic_reactor/utils/cachito.py:158]: It won't because `last_update_time` always gets set to `time.time()` on the first run because of `if last_updated_value`.

File: atomic_reactor/utils/cachito.py (hunk 4 above)
Comment (mprahl) [reply atomic_reactor/utils/cachito.py:154]: `last_update_time` is float of seconds but `response_json['updated']` is a string, so this won't work.

**Comment by athos-ribeiro** [comment]
Release notes added:

 - Cachito requests will now timeout based on the last update time

I will wait for @lcarva to ack his comments before merging (feel free to merge, Lui, RN already added).

**Comment by athos-ribeiro** [comment]
@lcarva, we decided to merge this so we can test it in time for the next release. If you think this should not have been merged, let me know and I will work on the missing bits :)
